### ISSUE TITLE
[pre-pull request] incompatibility between SQL Server syntax and interrogate()

### ISSUE THREAD

**Issue by pachadotdev** [rstudio/pointblank#144]
Hi Rich

Please check this out. I completed the next steps and I got to the point where SQL Server doesn't understand what dplyr is translating to SQL.

# 1: install fresh SQL Server (trial) locally

See https://docs.microsoft.com/en-us/sql/linux/quickstart-install-connect-ubuntu?view=sql-server-ver15

# 2: Create pointblank user as SA from dbeaver

```
CREATE LOGIN pointblank WITH PASSWORD = 'P01ntBl4nk';
CREATE USER pointblank FOR LOGIN pointblank ;
GRANT SELECT, INSERT, UPDATE, DELETE, EXECUTE, ALTER ON SCHEMA :: dbo TO pointblank;
```

# 3: Write nycflights13 to SQL Server

```
con <- DBI::dbConnect(
  odbc::odbc(),
  Driver = "SQLServer",
  Server = "127.0.0.1",
  UID = "pointblank",
  PWD = "P01ntBl4nk",
  Port = 1433
)

local_tables <- utils::data(package = "nycflights13")$results[, "Item"]

unique_index <- list(
  airlines = list("carrier"),
  planes = list("tailnum")
)

index <- list(
  airports = list("faa"),
  flights = list(
    c("year", "month", "day"), "carrier", "tailnum", "origin", "dest"
  ),
  weather = list(c("year", "month", "day"), "origin")
)

schema <- "dbo"

for (table in local_tables) {
  df <- getExportedValue("nycflights13", table)

  message("Creating table: ", table)

  table_name <- table

  DBI::dbWriteTable(
    con,
    table_name,
    df,
    unique_indexes = unique_index[[table]],
    indexes = index[[table]],
    temporary = FALSE,
    row.names = FALSE,
    overwrite = FALSE,
    append = TRUE
  )
}
```

# 4: Interrogate the database

Here the problem happens with both Microsoft ODBC Driver and RStudio Pro Drivers for SQL Server. See https://github.com/pachamaltese/pointblank/blob/master/tests/testthat/test-sqlserver.R#L22

```
dplyr::tbl(con, "planes") %>%
      create_agent(
        name = "master: 'planes' table",
        actions = al
      ) %>%
      col_vals_lte(vars(year), 2013L) %>%
      col_vals_gte(vars(year), 1956L) %>%
      col_vals_gt(vars(engines), 0L) %>%
      col_vals_gt(vars(seats), 0L) %>%
      col_vals_in_set(vars(manufacturer), "BOEING") %>%
      col_schema_match(
        schema = col_schema(
          manufacturer = "character",
          year = "integer",
          engines = "integer",
          seats = "integer"
        )
      ) %>%
      interrogate()
```

The interrogation returns this error:
```
── Interrogation Started - there are 4 steps ─────────────────────────────────────────────────
 Error: nanodbc/nanodbc.cpp:1617: 00000: [Microsoft][ODBC Driver 17 for SQL Server][SQL Server]Incorrect syntax near '<'.  [Microsoft][ODBC Driver 17 for SQL Server][SQL Server]Statement(s) could not be prepared. 
<SQL> 'SELECT "n"
FROM (SELECT COUNT(*) AS "n"
FROM (SELECT "tailnum", "year", "type", "manufacturer", "model", "engines", "seats", "speed", "engine", CASE
WHEN ((("pb_is_good_") IS NULL)) THEN (FALSE)
ELSE ("pb_is_good_")
END AS "pb_is_good_"
FROM (SELECT "tailnum", "year", "type", "manufacturer", "model", "engines", "seats", "speed", "engine", "year" <= 2013.0 AS "pb_is_good_"
FROM "planes") "dbplyr_017") "dbplyr_018") "dbplyr_019"' 
```

Which, in dbeaver can be tidied a bit to this query:
```
SELECT "n"
FROM (
	SELECT COUNT(*) AS "n"
	FROM (
		SELECT "tailnum", "year", "type", "manufacturer", "model", "engines", "seats", "speed", "engine",
		 CASE WHEN ((("pb_is_good_") IS NULL)) THEN (FALSE) ELSE ("pb_is_good_")
		 END AS "pb_is_good_"
		FROM (
			SELECT "tailnum", "year", "type", "manufacturer", "model", "engines", "seats", "speed", "engine", 
			"year" <= 2013.0 AS "pb_is_good_" FROM "planes") "dbplyr_017"
		 ) "dbplyr_018"
	 ) "dbplyr_019";
``` 

And here we can see that how the variable `pb_is_good_`  is created causes an SQL error:
```
SQL Error [102] [S0001]: Incorrect syntax near '<'.

org.jkiss.dbeaver.model.sql.DBSQLException: SQL Error [102] [S0001]: Incorrect syntax near '<'.
	at org.jkiss.dbeaver.model.impl.jdbc.exec.JDBCStatementImpl.executeStatement(JDBCStatementImpl.java:135)
	at org.jkiss.dbeaver.ui.editors.sql.execute.SQLQueryJob.executeStatement(SQLQueryJob.java:505)
	at org.jkiss.dbeaver.ui.editors.sql.execute.SQLQueryJob.lambda$0(SQLQueryJob.java:436)
	at org.jkiss.dbeaver.model.exec.DBExecUtils.tryExecuteRecover(DBExecUtils.java:170)
	at org.jkiss.dbeaver.ui.editors.sql.execute.SQLQueryJob.executeSingleQuery(SQLQueryJob.java:423)
	at org.jkiss.dbeaver.ui.editors.sql.execute.SQLQueryJob.extractData(SQLQueryJob.java:804)
	at org.jkiss.dbeaver.ui.editors.sql.SQLEditor$QueryResultsContainer.readData(SQLEditor.java:2996)
	at org.jkiss.dbeaver.ui.controls.resultset.ResultSetJobDataRead.lambda$0(ResultSetJobDataRead.java:120)
	at org.jkiss.dbeaver.model.exec.DBExecUtils.tryExecuteRecover(DBExecUtils.java:170)
	at org.jkiss.dbeaver.ui.controls.resultset.ResultSetJobDataRead.run(ResultSetJobDataRead.java:118)
	at org.jkiss.dbeaver.ui.controls.resultset.ResultSetViewer$ResultSetDataPumpJob.run(ResultSetViewer.java:4361)
	at org.jkiss.dbeaver.model.runtime.AbstractJob.run(AbstractJob.java:105)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:63)
Caused by: com.microsoft.sqlserver.jdbc.SQLServerException: Incorrect syntax near '<'.
	at com.microsoft.sqlserver.jdbc.SQLServerException.makeFromDatabaseError(SQLServerException.java:262)
	at com.microsoft.sqlserver.jdbc.SQLServerStatement.getNextResult(SQLServerStatement.java:1632)
	at com.microsoft.sqlserver.jdbc.SQLServerStatement.doExecuteStatement(SQLServerStatement.java:872)
	at com.microsoft.sqlserver.jdbc.SQLServerStatement$StmtExecCmd.doExecute(SQLServerStatement.java:767)
	at com.microsoft.sqlserver.jdbc.TDSCommand.execute(IOBuffer.java:7225)
	at com.microsoft.sqlserver.jdbc.SQLServerConnection.executeCommand(SQLServerConnection.java:3053)
	at com.microsoft.sqlserver.jdbc.SQLServerStatement.executeCommand(SQLServerStatement.java:247)
	at com.microsoft.sqlserver.jdbc.SQLServerStatement.executeStatement(SQLServerStatement.java:222)
	at com.microsoft.sqlserver.jdbc.SQLServerStatement.execute(SQLServerStatement.java:743)
	at org.jkiss.dbeaver.model.impl.jdbc.exec.JDBCStatementImpl.execute(JDBCStatementImpl.java:342)
	at org.jkiss.dbeaver.model.impl.jdbc.exec.JDBCStatementImpl.executeStatement(JDBCStatementImpl.java:132)
	... 12 more
```

# 5: Here's my session information

```
> sessionInfo()
R version 3.6.3 (2020-02-29)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 18.04.4 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas/libblas.so.3
LAPACK: /usr/lib/x86_64-linux-gnu/libopenblasp-r0.2.20.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=es_CL.UTF-8       
 [4] LC_COLLATE=en_US.UTF-8     LC_MONETARY=es_CL.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=es_CL.UTF-8       LC_NAME=C                  LC_ADDRESS=C              
[10] LC_TELEPHONE=C             LC_MEASUREMENT=es_CL.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] testthat_2.3.2        dplyr_1.0.0           pointblank_0.5.0.9000 dittodb_0.1.0        
[5] DBI_1.1.0            

loaded via a namespace (and not attached):
 [1] Rcpp_1.0.4.6      compiler_3.6.3    pillar_1.4.4      dbplyr_1.4.4      RPostgres_1.2.0  
 [6] tools_3.6.3       odbc_1.2.3        digest_0.6.25     packrat_0.5.0     bit_1.1-15.2     
[11] checkmate_2.0.0   gtable_0.3.0      RSQLite_2.2.0     memoise_1.1.0     lifecycle_0.2.0  
[16] tibble_3.0.1      pkgconfig_2.0.3   rlang_0.4.6       cli_2.0.2         rstudioapi_0.11  
[21] commonmark_1.7    yaml_2.2.1        RPostgreSQL_0.6-2 stringr_1.4.0     RMariaDB_1.0.9   
[26] sass_0.2.0        generics_0.0.2    vctrs_0.3.1       hms_0.5.3         grid_3.6.3       
[31] bit64_0.9-7       tidyselect_1.1.0  glue_1.4.1        R6_2.4.1          fansi_0.4.1      
[36] blastula_0.3.2    ggplot2_3.3.1     purrr_0.3.4       blob_1.2.1        magrittr_1.5     
[41] backports_1.1.8   scales_1.1.1      htmltools_0.5.0   ellipsis_0.3.1    gt_0.2.2         
[46] assertthat_0.2.1  colorspace_1.4-1  utf8_1.1.4        stringi_1.4.6     munsell_0.5.0    
[51] crayon_1.3.4 
```

**Comment by pachadotdev** [comment]
update: the same works in PostgreSQL and MariaDB/MySQL when I just edit the connection

**Comment by pachadotdev** [comment]
todo: maybe try this in the code
```
SELECT "tailnum", "year", "type", "manufacturer", "model", "engines", "seats", "speed", "engine",  CASE WHEN ("year" <= 2013.0) then(1) END AS "pb_is_good_" FROM "planes"
```

**Comment by rich-iannone** [comment]
Good idea. I can try that for sure. If `dplyr::case_when()` doesn't work for SQLite, another option is to use the CASE WHEN variant just for SQL Server.

**Comment by pachadotdev** [comment]
Let me try and I' ll send a decent PR

**Comment by rich-iannone** [comment]
I'm pretty excited about having some testing/bug-fixing for SQL Server being done. I wonder if the `col_schema_match()` validation step works at all. Perhaps it might be good to manually test all of the validation steps in isolation and see which ones clearly don't work?

**Comment by rich-iannone** [comment]
Closed by #146
